stl/inc/ximmortalize.h

```diff
@@ -0,0 +1,77 @@
+// ximmortalize.h internal header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _XIMMORTALIZE_H
+#define _XIMMORTALIZE_H
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+#ifdef _M_CEE_PURE
+template <class _Ty>
+union _Immortalizer_impl { // constructs _Ty, never destroys
+    _Immortalizer_impl() noexcept : _Storage{} {}
+    _Immortalizer_impl(const _Immortalizer_impl&) = delete;
+    _Immortalizer_impl& operator=(const _Immortalizer_impl&) = delete;
+    ~_Immortalizer_impl() {
+        // do nothing
+    }
+
+    _Ty _Storage;
+};
+
+#pragma warning(push)
+#pragma warning(disable : 4640) // construction of local static object is not thread-safe (/Wall)
+template <class _Ty>
+constexpr _Ty& _Immortalize() noexcept { // return a reference to an object that will live forever
+    /* MAGIC */ static _Immortalizer_impl<_Ty> _Static;
+    return _Static._Storage;
+}
+#pragma warning(pop)
+
+#else // ^^^ _M_CEE_PURE ^^^ // vvv !_M_CEE_PURE vvv
+template <class _Ty>
+union _Immortalizer_impl {

```

Comment:
Having pure and non-pure implementations of `_Immortalizer_impl` with the same name but different behavior seems like a recipe for ODR doom.

Reply:
pure isn't link compatible with non-pure so I don't think this matters.